src/algo/mod.rs

```diff
@@ -330,9 +329,7 @@ where
 
 #[derive(Copy, Clone, Debug)]
 struct NodeData {
-    index: Option<usize>,
-    lowlink: usize,
-    on_stack: bool,
+    rootindex: Option<usize>,
```

Comment:
Since you are optimizing for memory here, maybe use `Option<NonZeroUsize>`? Index will always be non-zero due to the invariant.